### ISSUE TITLE
(RFC) Can we cache CI artifacts to speed up CI cycle times...

### ISSUE THREAD

**Issue by grahamwhaley** [kata-containers/ci#64]
Each of our CI build/runs is pretty much a full isolated from source build, of almost all components. There are a couple of components it pulls as binary from an external source, but half the time the system builds those from source anyway due to version differences.
Given almost all components are identical across all builds, having the CI build everything from source is costing us a lot of CI resource time. We should discuss if there is any feasible way we can set up an 'artifact repository' that can somehow look at the version/commit/SHA of the source and the distro we are building on, and seeing if it can provide a pre-built binary...

There may be complications or wrinkles here, so here are some thought/notes:

- Some larger items we currently extract from our OBS binary builds if we can (kernel, qemu etc.),
    - but, if our version.yaml file does not match the pre-built binary (which happens often as we upgrade our `HEAD`, and the binaries are only from the last tagged release), then we build from source.
- Ensuring we match pre-built binaries not only to their correct 'versions' (which should be fairly easy using either a version number for some components such as kernel or qemu, or a git SHA for some other components) but - also match the correct distro and its update status, could be tricky.
- maybe just having a large [ccache](https://ccache.samba.org/) accessible by the build slaves would help greatly, and be a 'quick fix' for the qemu and kernel builds (they are in `C`, I don't think there is a `ccache` for `go`? - so it may not help with any k8s build/install for instance.)

For a start, we should probably have a first pass breakdown of where the time goes on a CI build at present. Before optimising, we should probably measure, measure, measure...

/cc @chavafg @jcvenegas @kata-containers/architecture-committee

**Comment by jodh-intel** [comment]
Speaking of measuring, this relates to https://github.com/kata-containers/ci/issues/15 since it's currently very hard to work out which part of the CI:

- is taking the longest.
- actually failed.

**Comment by jodh-intel** [comment]
ftr, I got burnt badly with caching and semaphore CI - stale caches / corrupt caches, and lots of time spent constructing valid cache data. I'm not against the idea but we'd need a solid plan and proof it's going to buy us something for the maintenance cost.

**Comment by grahamwhaley** [comment]
Understood @jodh-intel - I have vague memories of the ccache corruption etc.
From a very quick surf, I suspect the 'Jenkins way' to do this is to use a combination of [Maven](https://maven.apache.org/) and [Nexus](https://www.sonatype.com/nexus-repository-sonatype), along with Jenkins pipeline builds (like #15).
That all sounds like a fair slab of work.
If anybody has any experience of any of these - input here most welcome :-)

**Comment by grahamwhaley** [comment]
Out of interest, @cboylan - does Zuul have any notion or integrations to store build artifacts to speed up CI builds etc.?

**Comment by chavafg** [comment]
I took a look at the time that we spend building/installing artifacts for the CI systems.
The ones that take too long and it is where I think we can easily improve are `qemu` and `kernel`. Building the `kernel` takes around 8 minutes, while `qemu` takes 4 minutes.
In addition, building `CNI plugins` and `CRI-O` takes around 2 minutes and these ones don't change often, so they could also be candidates. 

Building the shim, proxy and runtime take less than 15 seconds total, so I think we are ok there. 
K8s and openshift are not being built, we took their releases binaries from their respective site, so we are not spending much time there.
